### PR TITLE
ICNet

### PR DESCRIPTION
**Pull request by KuangHaofei**
Adding ICNet model to model_zoo.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-1138-3 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1138/3/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/icnet.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+        scale_pred = F.contrib.BilinearResize2D(scale_pred4, height=h, width=w)
+        loss4 = super(ICNetLoss, self).hybrid_forward(F, scale_pred, label)
+
+        return loss1 + self.weights[0] * loss2 + \
+               self.weights[1] * loss3 + self.weights[2] * loss4
+
+    def hybrid_forward(self, F, *inputs, **kwargs):
+        """Compute loss"""
+        return self._weighted_forwarad(F, *inputs, **kwargs)
```
Comment (chongruo) [inline gluoncv/model_zoo/icnet.py]: typo (forward)

File: gluoncv/model_zoo/icnet.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        loss1 = super(ICNetLoss, self).hybrid_forward(F, scale_pred, label)
+
+        scale_pred = F.contrib.BilinearResize2D(scale_pred2, height=h, width=w)
+        loss2 = super(ICNetLoss, self).hybrid_forward(F, scale_pred, label)
+
+        scale_pred = F.contrib.BilinearResize2D(scale_pred3, height=h, width=w)
+        loss3 = super(ICNetLoss, self).hybrid_forward(F, scale_pred, label)
+
+        scale_pred = F.contrib.BilinearResize2D(scale_pred4, height=h, width=w)
+        loss4 = super(ICNetLoss, self).hybrid_forward(F, scale_pred, label)
```
Comment (chongruo) [inline gluoncv/model_zoo/icnet.py]: maybe it is better to go over the *input by "for loop"

File: gluoncv/model_zoo/icnet.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+        Hengshuang Zhao, Xiaojuan Qi, Xiaoyong Shen, Jianping Shi and Jiaya Jia.
+        "ICNet for Real-Time Semantic Segmentation on High-Resolution Images." *ECCV*, 2018
+
+    Examples
+    --------
+    >>> model = ICNet(nclass=19, backbone='resnet50')
+    >>> print(model)
+    """
+    # pylint: disable=arguments-differ
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: I think you already disable arguments-differ in the beginning of the file, there is no need to disable it again. Please remove this line.

File: gluoncv/model_zoo/icnet.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+    Parameters
+    ----------
+    nclass : int
+        Number of categories for the training dataset.
+    backbone : string
+        Pre-trained dilated backbone network type (default:'resnet50'; 'resnet50',
+        'resnet101' or 'resnet152').
+    norm_layer : object
+        Normalization layer used in backbone network (default: :class:`mxnet.gluon.nn.BatchNorm`;
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: No matching parentheses, and no need to add ;. Change it to `(default: :class: mxnet.gluon.nn.BatchNorm)`

File: gluoncv/model_zoo/icnet.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+            self.conv_sub2 = ConvBnRelu(512, 256, 1)
+            self.conv_sub2.initialize(ctx=ctx)
+            self.conv_sub2.collect_params().setattr('lr_mult', 10)
+
+    def hybrid_forward(self, F, x):
+        # sub 1
+        x_sub1_out = self.conv_sub1(x)
+
+        # sub_2
+        x_sub2 = F.contrib.BilinearResize2D(x, height=x.shape[2] // 2, width=x.shape[3] // 2)
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: In order to use hybrid_forward, you cannot use `.shape` here. Please address it using something like `self._up_kwargs['height'] ` to pass in the shape.

File: gluoncv/model_zoo/icnet.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+        x = self.conv1(x_sub2)
+        x = self.bn1(x)
+        x = self.relu(x)
+        x = self.maxpool(x)
+
+        x = self.layer1(x)
+        x_sub2_out = self.layer2(x)
+
+        # sub 4
+        x_sub4 = F.contrib.BilinearResize2D(x_sub2_out,
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: Same thing here, can't get shape inside hybrid_forward.

File: gluoncv/model_zoo/icnet.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+            with self.conv_hign.name_scope():
+                self.conv_hign.add(nn.Conv2D(in_channels=high_channels, channels=out_channels,
+                                             kernel_size=1, use_bias=False))
+                self.conv_hign.add(norm_layer(in_channels=out_channels))
+
+            self.conv_low_cls = nn.Conv2D(in_channels=out_channels, channels=nclass,
+                                          kernel_size=1, use_bias=False)
+
+    def hybrid_forward(self, F, x_low, x_high):
+        x_low = F.contrib.BilinearResize2D(x_low, height=x_high.shape[2], width=x_high.shape[3])
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: same thing here, pass in the shape instead of getting a shape inside hybrid_forward.

File: gluoncv/model_zoo/icnet.py [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+    def __init__(self, nclass, norm_layer=nn.BatchNorm):
+        super(_ICHead, self).__init__()
+        self.cff_12 = CascadeFeatureFusion(128, 64, 128, nclass, norm_layer)
+        self.cff_24 = CascadeFeatureFusion(256, 256, 128, nclass, norm_layer)
+
+        with self.name_scope():
+            self.conv_cls = nn.Conv2D(in_channels=128, channels=nclass,
+                                      kernel_size=1, use_bias=False)
+
+    # pylint: disable=arguments-differ
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: Remove this pylint disable.

File: gluoncv/model_zoo/icnet.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    def hybrid_forward(self, F, x_sub1, x_sub2, x_sub4):
+        outputs = []
+
+        x_cff_24, x_24_cls = self.cff_24(x_sub4, x_sub2)
+        outputs.append(x_24_cls)
+        x_cff_12, x_12_cls = self.cff_12(x_cff_24, x_sub1)
+        outputs.append(x_12_cls)
+
+        up_x2 = F.contrib.BilinearResize2D(x_cff_12,
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: same thing here, change get shape.

File: gluoncv/model_zoo/icnet.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+        return tuple(outputs)
+
+
+def get_icnet(dataset='citys', backbone='resnet50', pretrained=False,
+              root='~/.mxnet/models', pretrained_base=True, ctx=cpu(0), **kwargs):
+    r"""Image Cascade Network
+    Parameters
+    ----------
+    dataset : str, default citys
+        The dataset that model pretrained on. (cityscapes)
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: change `cityscapes` to `default: cityscapes`

File: gluoncv/model_zoo/icnet.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+from mxnet.gluon.nn import HybridBlock
+from gluoncv.model_zoo.segbase import SegBaseModel
+from gluoncv.model_zoo.pspnet import _PSPHead
+from gluoncv.loss import SoftmaxCrossEntropyLoss
+
+
+__all__ = ['ICNet', 'get_icnet', 'get_icnet_resnet50_citys', 'ICNetLoss']
+
+
+class ICNetLoss(SoftmaxCrossEntropyLoss):
```
Comment (bryanyzhu) [inline gluoncv/model_zoo/icnet.py]: I think this is a loss function, which should be put into this file, https://github.com/dmlc/gluon-cv/blob/master/gluoncv/loss.py

File: gluoncv/model_zoo/icnet.py (hunk 1 above)
Comment (KuangHaofei) [reply gluoncv/model_zoo/icnet.py]: OK, thank you for your reminder!

**Comment by mli** [comment]
Job PR-1138-12 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1138/12/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1138-14 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1138/14/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-1138-16 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-1138/16/index.html 
Code coverage of this PR:  vs. Master:

**Comment by bryanyzhu** [comment]
@zhanghang1989 Please help review this PR. This PR adds ICNet, performance outperforms original paper. Model zoo and test coverage are provided. 

Since ICNet is a real-time framework, it only requires single-scale testing. Hence, the changes to `test.py` is a bit more. But we use if else control flow so shouldn't introduce any code breaking. I also tested several models, seems fine.
